### ISSUE TITLE
Last Will is never treated as retained

### ISSUE THREAD

**Issue by knr1** [moquette-io/moquette#171]
We noticed that moquette does not seem to retain the last will published by moquette, even though the message has been prepared this way during connection.

However, according to the specification OASIS 3.1.1, 
"If the Will Flag is set to 1, the Will QoS and Will Retain fields in the Connect Flags will be used by the Server, and the Will Topic and Will Message fields MUST be present in the payload [MQTT-3.1.2-9]."

In order to provoke this behaviour, we have done the folowing:
Create a connection with a last will: 
(clientid "a", qos "1", retain "true", topic "status", message "offline")
when the connection is disconnected forcefully (by killing the process), the broker sends the will message.
We checked this with a client via WebSocket: (clientId "b", subscribe "#", qos "1")

But when connecting and checking with a client c some moments after client b has confirmed the last will, the message is not retained, hence cannot be seen by c (or any other client) anymore....

Could you elaborate on that?

**Comment by andsel** [comment]
Hi @knr1 you are right and Moquette implement it. I've just added an integration test to check your use case (commit: dbe9c7bcaa179a4c4b5ca0633862b12e09f1a6eb) could you provide the log of your real case? (also witch version on Moquette you are using, if snapshot, try to update it, which OS, possibly a unit test or code to make it repeatable).

**Comment by knr1** [comment]
Sorry for the long absence... Here I am again.
For me, the situation is still the same even with the newest checkout of the 0.9 broker.

I started the broker... here is its printout:

---

Building Moquette - broker 0.9-SNAPSHOT

---

Downloading: https://repo.eclipse.org/content/repositories/paho-releases/io/moquette/moquette-netty-parser/0.9-SNAPSHOT/maven-metadata.xml
Downloading: https://oss.sonatype.org/content/repositories/snapshots/io/moquette/moquette-netty-parser/0.9-SNAPSHOT/maven-metadata.xml

Downloading: https://repo.eclipse.org/content/repositories/paho-releases/io/moquette/moquette-parent/0.9-SNAPSHOT/maven-metadata.xml
Downloading: https://oss.sonatype.org/content/repositories/snapshots/io/moquette/moquette-parent/0.9-SNAPSHOT/maven-metadata.xml

--- exec-maven-plugin:1.2.1:exec (default-cli) @ moquette-broker ---
log4j:WARN No appenders could be found for logger (io.moquette.server.Server).
log4j:WARN Please initialize the log4j system properly.
log4j:WARN See http://logging.apache.org/log4j/1.2/faq.html#noconfig for more info.
Server started, version 0.9-SNAPSHOT

Here my test setup with which you can provoke the ?strange? behaviour:

First I create a 'Service' with a last will:

```
public class TestService {
    public TestService() throws MqttException {
        IMqttAsyncClient mqttClient = new MqttAsyncClient("tcp://localhost:1883", "TestServiceClientID");
        MqttConnectOptions options = new MqttConnectOptions();
        options.setWill("will", "offline".getBytes(), 1, true);
        mqttClient.connect(options).waitForCompletion();
        mqttClient.publish("will", "online".getBytes(), 1, true);
    }
    public static void main(String[] args) throws MqttException, IOException {
        new TestService();
        System.in.read();
     }
}
```

Then a little 'Agent' listening to all Topics:

```
public class TestAgent implements MqttCallback {
    public TestAgent() throws MqttException {
        IMqttAsyncClient mqttClient = new MqttAsyncClient("tcp://localhost:1883", "TestAgentClientID");
        MqttConnectOptions options = new MqttConnectOptions();
        options.setCleanSession(false);
        mqttClient.setCallback(this);
        mqttClient.connect(options).waitForCompletion();
        mqttClient.subscribe("#", 1);
    }
    @Override
    public void connectionLost(Throwable thrwbl) {
    }
    @Override
    public void messageArrived(String string, MqttMessage mm) throws Exception {
        System.out.printf("Will: %s, Message: %s", string, new String(mm.getPayload()));
    }
    @Override
    public void deliveryComplete(IMqttDeliveryToken imdt) {
    }
    public static void main(String[] args) throws MqttException, IOException {
        new TestAgent();
        System.in.read();
    }
}
```

Now start the Agent
then start the Service
What the Agent prints out is... that the Service is 'online'.

Now forcefully stop the Service...
What the Agent prints out is... that the Service is 'offline'

Now comes the case: Forcefully stop the Agent.
Restart the Agent.
What the Agent prints out is... that the Service is 'online'

However, the service has not been started any-more... So what could explain this ?strange? behaviour?

**Comment by andsel** [comment]
Fixed by PR #343
